src/ruby/puppetserver-lib/puppet/server/http_client.rb

```diff
@@ -96,7 +90,11 @@ def create_common_request_options(url, headers, options)
 
     # Ensure multiple requests are not made on the same connection
     headers["Connection"] = "close"
-    request_options = RequestOptions.new(build_url(url))
+
+    # Java can't handle the Ruby URI type
+    url = url.to_s if url.is_a?(URI)
+    # TODO Add params to this URL
```

Comment:
??

Reply:
I'm not actually sure what we're supposed to do with the query params now, I need to look into it a little more. The old API didn't have a separate arg for them, so I assumed they used to be just part of the URL if you supplied them. However, I also suspect that nothing that uses this client actually does supply them, so it might coincidentally not matter right now...

Reply:
It would be nice to clarify that.

Reply:
The old API accepted query parameters in the [`path` parameter](https://github.com/puppetlabs/puppet/blob/f638777d6ece022dac8f8c2f6313870d494fc4aa/lib/puppet/network/http/connection.rb#L92) which could either be `http://.../a?b=c` or `/a?b=c`. It just "kind of worked" because `Net::HTTP` accepts either form, yay ruby! The new API makes it possible to specify the URL and params separately as that's the common use case in puppet, e.g. we have a [`Puppet[:reporturl]` and want to pass along some params](https://github.com/puppetlabs/puppet/blob/f638777d6ece022dac8f8c2f6313870d494fc4aa/lib/puppet/reports/http.rb#L17-L26). If you include explicit query parameters then they [overwrite the query in the URL](https://github.com/puppetlabs/puppet/blob/f638777d6ece022dac8f8c2f6313870d494fc4aa/lib/puppet/http/client.rb#L283). I thought about making that an error or warning if they were in both places, but didn't get around to doing that.

Reply:
I think what I've done here will be same as that.